### PR TITLE
DSD-1035: Refactoring Header functions

### PR DESCRIPTION
**Pull request by EdwinGuzman**
Fixes JIRA ticket [DSD-1035](https://jira.nypl.org/browse/DSD-1035)

## This PR does the following:

- Refactors functions that are used within children components in the `Header` components.
- Adds a new `useCloseDropDown` hook to close dropdown-like components by mouse clicks outside of it and by pressing the escape key.
- Creates a `PatronContext` and `PatronProvider` to manage the patron data to remove prop drilling.

## How has this been tested?

Locally in Storybook. Note: because of the iframe in Storybook, it's hard to test clicking outside of a component to close a dropdown. The clicks need to be inside of the `Header` component and not past its iframe where it is rendered, otherwise, those event listeners won't work. This is only an issue in Storybook.

## Accessibility concerns or updates

- N/A

### Checklist:

- [ ] I have updated the Storybook documentation accordingly.
- [ ] I have added relevant accessibility documentation for this pull request.
- [ ] All new and existing tests passed.

### Front End Review:

- [ ] Review the Vercel preview deployment once it is ready.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -19,6 +19,7 @@ Currently, this repo is in Prerelease. When it is released, this project will ad
 - Refactors `SearchButton` and `SearchForm` to implement the mobile search form.
 - Adds the mobile navigation menu for the `Header` component with `MobileNav` and `MobileNavButton` components.
 - Adds the mobile login menu for the `Header` component.
+- Adds the `useCloseDropDown` hook to manage closing dropdown-like components through clicks outside of it and by pressing the escape key.
```
Comment (oliviawongnyc) [inline CHANGELOG.md:22]: Should we add something here about patronContext?

File: src/components/Header/context/patronContext.test.tsx [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+import { PatronContext, PatronProvider } from "./patronContext";
+
+describe("PatronContext", () => {
+  const PatronTestComponent = () => {
+    const name = React.useContext(PatronContext);
+    return <p>{name}</p>;
+  };
+
+  it("passes down the patron name value down to its children", () => {
```
Comment (oliviawongnyc) [inline src/components/Header/context/patronContext.test.tsx:29]: Maybe overkill, but should we test when the patronName is not provided, or is an empty string?

File: src/hooks/__tests__/useCloseDropDown.test.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      screen.queryByText("The drop down is open.")
+    ).not.toBeInTheDocument();
+
+    fireEvent.click(screen.getByText("Toggle"));
+
+    expect(screen.getByText("The drop down is open.")).toBeInTheDocument();
+
+    // Clicking outside of the dropdown should close it. We mock
+    // this by "clicking" on any element outside of the ref.
+    fireEvent.mouseDown(element);
```
Comment (oliviawongnyc) [inline src/hooks/__tests__/useCloseDropDown.test.tsx:40]: I am guessing you can't use fireEvent.click on a p element?

File: src/hooks/__tests__/useCloseDropDown.test.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    expect(
+      screen.queryByText("The drop down is open.")
+    ).not.toBeInTheDocument();
+
+    fireEvent.click(screen.getByText("Toggle"));
+
+    expect(screen.getByText("The drop down is open.")).toBeInTheDocument();
+
+    // Pressing the esc key should close the dropdown.
+    fireEvent.keyDown(screen.getByText("Toggle"), { key: "Escape" });
```
Comment (oliviawongnyc) [inline src/hooks/__tests__/useCloseDropDown.test.tsx:63]: I did some research and learned that you need to include a domNode in fireEvent.keyDown (hence why you added screen.getByText("Toggle")), but maybe we can add that as a comment?

File: src/hooks/useCloseDropDown.stories.mdx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+the developer to pass a function that accepts a boolean value that updates the
+state of the component where the hook is called. This callback function is
+triggered when the user clicks outside of the component, or presses the escape
+key.
+
+The `ref` argument is used in Chakra's `useOutsideClick` function to "close" a
+component. The `ref` allows the function to know where "outside the component"
+is.
+
+In the example below, the `ButtonDropDownExample` component keeps track of it's
```
Comment (oliviawongnyc) [inline src/hooks/useCloseDropDown.stories.mdx]: Nit: "its" should not have an apostrophe.

**Review by oliviawongnyc — COMMENTED** [review]
Loved seeing these improvements. Just a couple of small notes!

File: CHANGELOG.md (hunk 1 above)
Comment (EdwinGuzman) [reply CHANGELOG.md:22]: Will add.

File: src/components/Header/context/patronContext.test.tsx (hunk 2 above)
Comment (EdwinGuzman) [reply src/components/Header/context/patronContext.test.tsx:29]: Nope, that's a solid update.

File: src/hooks/__tests__/useCloseDropDown.test.tsx (hunk 3 above)
Comment (EdwinGuzman) [reply src/hooks/__tests__/useCloseDropDown.test.tsx:40]: Yes. I'll write that.

File: src/hooks/__tests__/useCloseDropDown.test.tsx (hunk 4 above)
Comment (EdwinGuzman) [reply src/hooks/__tests__/useCloseDropDown.test.tsx:63]: Yes, will do.

**Comment by EdwinGuzman** [comment]
Updated.

File: src/hooks/useCloseDropDown.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import { useEffect } from "react";
+import { useOutsideClick } from "@chakra-ui/react";
+
+/**
+ * Close dropdowns when clicking outside of them or by pressing
+ * the escape key. It expects an action callback that will set
+ * the dropdown state to false (close).
+ */
+export const useCloseDropDown = (
+  actionCb: (val: boolean) => void,
```
Comment (bigfishdesign13) [inline src/hooks/useCloseDropDown.ts:10]: Why use `actionCb` and not just `cb` or `callback`?

**Review by bigfishdesign13 — APPROVED** [review]
This looks good.  I just have one comment with one simple question.  

Although, I am also curious, can a hook handing the opposite action (useOpenDropDown) make sense?  Looking at the code example in the story, it looks like the necessary code when going in the open direction is not as complicated as the code for closing.

File: src/hooks/useCloseDropDown.ts (hunk 6 above)
Comment (EdwinGuzman) [reply src/hooks/useCloseDropDown.ts:10]: I think a more descriptive variable name is better. It also expects it to be an actionable callback and not just any callback, so hopefully that's clear.

**Comment by EdwinGuzman** [comment]
Creating a hook like `useOpenDropDown` does make sense. I can investigate and see if it'll make the code easier.

**Pull request merged**